### PR TITLE
merge shard series set

### PR DESCRIPTION
**Pull request by u-veles-a**
## Description
- add mergeShardSeriesSet - merger for SeriesSet from shards.
- add mergeShardChunkSeriesSet - merger for ChunkSeriesSet from shards.

### CODE REVIEW

File: pp/go/storage/querier/merge_series_set.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+)
+
+// mergeShardSeriesSet merges many [storage.SeriesSet] together from different shards.
+// Since there are always different series in the shards, there is no need to check them for Equal.
+type mergeShardSeriesSet struct {
+	heap       seriesSetHeap
+	currentSet storage.SeriesSet
+}
+
+// NewMergeShardSeriesSet returns a new [storage.SeriesSet] that merges many [storage.SeriesSet] together.
```
Comment (vporoshok) [inline pp/go/storage/querier/merge_series_set.go]: Since this helper is exported and accepts the generic `storage.SeriesSet` interface, please make the narrower contract explicit here. The implementation assumes only our shard-local sets are passed in: no `nil`, no `Err()`/`Warnings()`, sorted output, and no identical label sets across shards. That is a reasonable optimization for the current caller, but without documenting it a future caller may treat this as a drop-in replacement for `storage.NewMergeSeriesSet` and silently lose child iterator errors or duplicate-label merging. The same applies to `NewMergeShardChunkSeriesSet` below.

File: pp/go/storage/querier/merge_series_set_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/pp/go/cppbridge"
+	"github.com/prometheus/prometheus/pp/go/model"
+	"github.com/prometheus/prometheus/pp/go/storage/head/shard"
+	"github.com/prometheus/prometheus/pp/go/storage/querier"
+	"github.com/prometheus/prometheus/pp/go/storage/storagetest"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func BenchmarkMergeSeriesSet(b *testing.B) {
```
Comment (vporoshok) [inline pp/go/storage/querier/merge_series_set_test.go:163]: This benchmark is useful, but this PR also changes query correctness behavior by replacing the generic Prometheus merger in instant, range, and chunk selects. Please add a small correctness test suite for the optimized merger: sorted interleaving across shards, empty shards, the chunk-series path, and the unsupported duplicate-label case or its documented invariant. That would lock in the assumptions this faster merger relies on.

**Review by vporoshok — COMMENTED** [review]
Thanks, the optimization makes sense for the shard-local query path. I do not think the error propagation point should block this PR if these helpers are intentionally limited to the in-tree shard SeriesSet implementations, which currently do not produce errors or warnings.

I left two non-blocking comments: one to make that narrower contract explicit, and one to add a small correctness test suite around the optimized merger so future changes do not accidentally treat it as a drop-in replacement for Prometheus' generic merger.

**Review by vporoshok — APPROVED** [review]
